kvdb/etcd/stm.go

```diff
@@ -732,6 +942,72 @@ func (s *stm) OnCommit(cb func()) {
 	s.onCommit = cb
 }
 
+// Prefetch will prefetch the passed keys and prefixes in one transaction.
+// Keys and prefixes that we already have will be skipped.
+func (s *stm) Prefetch(keys []string, prefixes []string) {
```

Comment:
I must say that I am a little intimidated by the STM code. It is almost as if logic that is normally part of a (battle-tested) database server is now implemented in `lnd` itself. Do you think it is solid enough to trust it with critical Lightning data? It feels risky, but maybe I am not familiar enough with etcd.

Reply:
Yeah since we needed to bend the usual implementation to fit our pattern/interfaces, we ended up having to write more custom logic here than originally anticipated. I also get somewhat squeamish when we need to make significant changes in this area. FWIW, if we want to rely on them, there're [tools like Jepsen](https://jepsen.io/) that can be used to find consistency violating issues in distributed databases. A run of the tool a few years ago actually found a number of issues in the `etcd` logic as well as some of the concurrency APIs (leasing, locking, etc).